Parse "lang" member of Web Application Manifest

christianliebel opened the pull request:
#### e95d02b6d9505be0cc0507109990fd3a56240321
<pre>
Parse &quot;lang&quot; member of Web Application Manifest
<a href="https://bugs.webkit.org/show_bug.cgi?id=278984">https://bugs.webkit.org/show_bug.cgi?id=278984</a>

Reviewed by Anne van Kesteren, Darin Adler, and Yusuke Suzuki.

Implements the &quot;lang&quot; member to the web application manifest parser per spec:
<a href="https://www.w3.org/TR/appmanifest/#lang-member">https://www.w3.org/TR/appmanifest/#lang-member</a>
Along with the required encoders/decoders and API tests.

Adds a helper function to IntlObject to canonicalize language tags without the overhead of UTF-8 conversion.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::canonicalizeUnicodeLocaleID):
* Source/JavaScriptCore/runtime/IntlObject.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parseManifest):
(WebCore::ApplicationManifestParser::parseLang):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(-[_WKApplicationManifest initWithCoder:]):
(-[_WKApplicationManifest encodeWithCoder:]):
(-[_WKApplicationManifest lang]):
* Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
(ApplicationManifestParserTest::testLang):
(ApplicationManifestParserTest::testId):
(TEST_F(ApplicationManifestParserTest, Lang)):

Canonical link: <a href="https://commits.webkit.org/307465@main">https://commits.webkit.org/307465@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3c542199f880b92dd8b71ee4d124ffe99448f668

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/144514 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/17193 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/8754 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/153184 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/98149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/146389 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/17675 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/17087 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/111124 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/98149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/147477 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/13513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/129778 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/92036 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/12900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/10656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/630 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/136505 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/122406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/6468 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/155497 "Built successfully") | 
| [✅ 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/5323 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/17045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/7524 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/119126 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/17083 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/14271 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/119482 "Passed tests") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/15313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/127681 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/72586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/22287 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/16667 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/6084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/175802 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/16403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/80446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/45280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/16612 "Build is in progress. Recent messages:") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/16467 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

marcoscaceres commented on Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
```diff
@@ -522,6 +527,28 @@ TEST_F(ApplicationManifestParserTest, Orientation)
+    testLang("\"en-x-custom\""_s, "en-x-custom"_s);
+
+    // Language tags should be canonicalized.
+    testLang("\"DE-DE\""_s, "de-DE"_s);
```
Maybe just add the following to canonicalize and trim: 
```suggestion
    testLang("\"  DE-DE  \""_s, "de-DE"_s);
```

annevk commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -261,6 +263,20 @@ const std::optional<ScreenOrientationLockType> ApplicationManifestParser::parseO
 
+String ApplicationManifestParser::parseLang(const JSON::Object& manifest)
+{
+    String lang = parseGenericString(manifest, "lang"_s);
```
auto

darinadler commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -261,6 +263,20 @@ const std::optional<ScreenOrientationLockType> ApplicationManifestParser::parseO
+        return { };
+    }
+
+    return JSC::canonicalizeUnicodeLocaleID(lang.utf8());
```
Converting to UTF-8 on the way in is a peculiar step in canonicalizing this ID, and even an unnecessarily inefficient one since a structurally valid language tag is guaranteed not to have non-ASCII characters and likely is almost always already stored as an 8-bit string, while the `utf8` function always returns newly allocated memory than then has to be deallocated. Building this inefficient idiom into the interface between JavaScriptCore and WebCore isn’t great, even for something trivial like this. The conversion step is required only because the existing function from `IntlObject.h` happens to take a `CString`. I think we should have a function that takes a `String` or `StringView` that does the same operation that we export from `IntlObject.h`, even if for now it has creates a `CString` internally for reasons of expediency; it would be easy to write a version that doesn’t unnecessarily allocate and deallocate memory for an extra copy of the string in the common case, though.

annevk replied:
@christianliebel can you create a wrapper function in `IntlObject.h` to address this comment?

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
```diff
@@ -203,6 +203,11 @@ class ApplicationManifestParserTest : public testing::Test {
 
+    void testLang(const String& rawJSON, const String& expectedValue)
+    {
+        EXPECT_EQ(expectedValue, parseTopLevelProperty("lang"_s, rawJSON).lang);
```
I don’t understand why we are adding the parsing, but it has no effect on the web platform. It’s very strange that the only tests are in C++. Is there some web-platform-visible effect of the parsing that can be tested more end to end?

christianliebel replied:
This change extends the Web Application Manifest, which is evaluated when a user adds a website to the home screen. Safari uses these values to determine the icon, suggested name, etc., for the web clip. After parsing, all values are passed to the platform-specific side, so there are indeed no web-platform-visible effects.

darinadler replied:
Are you contributing a pull request that parses more and hoping to influence the Safari developers to make use of the new values?

marcoscaceres replied:
Yeah, it's more for standards compliance. Christian is a co-editor with me on the Manifest spec. On the W3C side, we require implementation before we add things to the spec. At the same time, I'm coordinating with Safari folks and letting them know of the changes - but it's up to them if they actually adopt them (i.e., that's a product decision for Safari, with the usual disclaimer that "Apple doesn't comment on future features and products"). Should Safari choose to use these values, they are there.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
```diff
@@ -74,6 +74,7 @@ WK_CLASS_AVAILABLE(macos(10.13.4), ios(11.3))
 
 @property (nonatomic, readonly, nullable, copy) NSString *rawJSON;
 @property (nonatomic, readonly) _WKApplicationManifestDirection dir;
+@property (nonatomic, readonly, nullable, copy) NSString *lang;
```
I need to understand the big picture. Why are we adding all these getters to this Objective-C class. What is this SPI for? Is this a building block for some feature we are working on?

marcoscaceres replied:
This is for create the UI that Safari uses when presenting the Add to Home Screen/Dock UI. The larger picture is to provide language hints, if needed, which would only have an impact potentially accessibility: the name of the application being read correctly in some language, for instance.

Constellation reviewed (approved):
r=me on JSC change

annevk commented on Source/JavaScriptCore/runtime/IntlObject.cpp:
```diff
@@ -731,6 +731,14 @@ String canonicalizeUnicodeLocaleID(const CString& tag)
+String canonicalizeUnicodeLocaleID(const StringView tag)
+{
+    if (tag.containsOnlyASCII())
+        return canonicalizeUnicodeLocaleID(CString(tag.span8()));
```
This is wrong. A string that is only ASCII is not guaranteed to use a 8-bit buffer. You need to use is8bit or some such. And perhaps you need to use both in this case?

annevk reviewed (commented):
We should add some tests for non-ASCII. We want just the ASCII range, just the ISO-8859-1 range, and beyond.

annevk commented on Source/JavaScriptCore/runtime/IntlObject.cpp:
```diff
@@ -731,6 +731,17 @@ String canonicalizeUnicodeLocaleID(const CString& tag)
+{
+    if (tag.is8Bit() && tag.containsOnlyASCII())
+        return canonicalizeUnicodeLocaleID(CString(tag.span8()));
+
+    // BCP 47 language tags should be ASCII-only.
+    ASSERT_NOT_REACHED();
+    return canonicalizeUnicodeLocaleID(tag.utf8());
+}
```
```suggestion
String canonicalizeUnicodeLocaleID(const StringView tag)
{
    ASSERT(tag.containsOnlyASCII());
    return canonicalizeUnicodeLocaleID(tag.utf8());
}
```
This is wrong for the same reason the old code was wrong. Even if it's all ASCII, you're not guaranteed to hit `is8Bit()`. You can have a non-8bit string that's all ASCII and we wouldn't want to assert on them.

Let's just make this a simple wrapper for now and improve performance once it shows up in benchmarks.

annevk commented on Source/JavaScriptCore/runtime/IntlObject.h:
```diff
@@ -154,8 +154,9 @@ bool isUnicodeScriptSubtag(StringView);
-bool isStructurallyValidLanguageTag(StringView);
-String canonicalizeUnicodeLocaleID(const CString& languageTag);
+JS_EXPORT_PRIVATE bool isStructurallyValidLanguageTag(StringView);
+JS_EXPORT_PRIVATE String canonicalizeUnicodeLocaleID(const CString& languageTag);
```
We don’t want to export this one? Missed this before, sorry

marcoscaceres replied:
If I remove them I get a linker error as it can't find the method(s).

annevk replied:
I’m not saying all, just this CString one.

christianliebel committed:
Parse "lang" member of Web Application Manifest
https://bugs.webkit.org/show_bug.cgi?id=278984

Reviewed by Anne van Kesteren, Darin Adler, and Yusuke Suzuki.

Implements the "lang" member to the web application manifest parser per spec:
https://www.w3.org/TR/appmanifest/#lang-member
Along with the required encoders/decoders and API tests.

Adds a helper function to IntlObject to canonicalize language tags without the overhead of UTF-8 conversion.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::canonicalizeUnicodeLocaleID):
* Source/JavaScriptCore/runtime/IntlObject.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parseManifest):
(WebCore::ApplicationManifestParser::parseLang):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(-[_WKApplicationManifest initWithCoder:]):
(-[_WKApplicationManifest encodeWithCoder:]):
(-[_WKApplicationManifest lang]):
* Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
(ApplicationManifestParserTest::testLang):
(ApplicationManifestParserTest::testId):
(TEST_F(ApplicationManifestParserTest, Lang)):

Canonical link: https://commits.webkit.org/307465@main
